lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp

```diff
@@ -0,0 +1,178 @@
+/// inference support.
+// TODO: move into ArithBuilder once ops have been moved into arith.
+struct AffineBuilder {
+  AffineBuilder(OpBuilder &b, Location loc) : b(b), loc(loc) {}
+
+  Value add(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e + rhs.e}, ValueRange{lhs, rhs});
+  }
+  Value sub(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e - rhs.e}, ValueRange{lhs, rhs});
+  }
+  Value mul(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e * rhs.e}, ValueRange{lhs, rhs});
+  }
+  Value ceil(Tie lhs, Tie rhs) {
+    return b.createOrFold<AffineApplyOp>(
+        loc, ArrayRef<AffineExpr>{lhs.e.ceilDiv(rhs.e)}, ValueRange{lhs, rhs});
+  }
+  Value min(ValueRange vals) {
+    return b.createOrFold<AffineMinOp>(
+        loc, AffineMap::getMultiDimIdentityMap(vals.size(), b.getContext()),
+        vals);
+  }
+  Value max(ValueRange vals) {
+    return b.createOrFold<AffineMinOp>(
+        loc, AffineMap::getMultiDimIdentityMap(vals.size(), b.getContext()),
+        vals);
+  }
+
+ private:
+  OpBuilder &b;
+  Location loc;
+};
+
+struct TileOpToInParallelRewriter
+    : public OpRewritePattern<linalg_ext::TileOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg_ext::TileOp tileOp,
+                                PatternRewriter &rewriter) const override {
+    // TODO: verifier.
+    assert(tileOp.getNumResults() > 0 &&
+           tileOp.outs().size() == tileOp.getNumResults());
+
+    // TODO: when supported, iterate over the tensor of sizes. This will be
+    // iterating through a level of indirection.
+
+    // Construct the loop bounds based on the canonical arithmetic progression.
+    Location loc = tileOp.getLoc();
+    Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value totalSize =
+        rewriter.create<tensor::DimOp>(loc, tileOp.outs().front(), zero);
+    Value step = tileOp.tile_sizes();
+    assert(step.getType().isa<IndexType>() && "NYI: not an index type");
+
+    AffineBuilder ab(rewriter, loc);
+    AffineExpr i, j, M;
+    bindDims(rewriter.getContext(), i, j);
+    bindSymbols(rewriter.getContext(), M);
+    Value sizePerThread = ab.ceil(Tie{i, totalSize}, Tie{M, step});
+
+    // Construct the op without a body builder: we need to clone the ops in the
+    // body explicitly after having access to the new bbArgs.
+    // As a consequence, `ensureTerminator` is not called and the body has no
+    // terminator.
+    linalg_ext::InParallelOp inParallelOp =
+        rewriter.create<linalg_ext::InParallelOp>(loc, tileOp->getResultTypes(),
+                                                  sizePerThread);
+    inParallelOp.ensureTerminator(inParallelOp.region(), rewriter, loc);
```

Comment:
Do we still need this? I think that with the new bodyless builder this should be unnecessary.